core/src/main/scala/kafka/raft/KafkaRequestPurgatory.scala

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.raft
+
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.atomic.AtomicBoolean
+
+import kafka.server.{DelayedOperation, DelayedOperationPurgatory}
+import kafka.utils.Logging
+import kafka.utils.timer.Timer
+import org.apache.kafka.common.errors.TimeoutException
+import org.apache.kafka.raft.RequestPurgatory
+
+/**
+ * Simple purgatory shim for integration with the Raft library. We assume that
+ * both [[await()]] and [[completeAll()]] are called in the same thread.
+ */
+class KafkaRequestPurgatory(brokerId: Int, timer: Timer, reaperEnabled: Boolean = true)
+  extends RequestPurgatory[Unit] with Logging {
+
+  private val key = new Object()
+  private val purgatory = new  DelayedOperationPurgatory[DelayedRaftRequest](
+    "raft-request-purgatory", timer, brokerId, reaperEnabled = reaperEnabled)
+
+  override def await(future: CompletableFuture[Unit], maxWaitTimeMs: Long): Unit = {
+    val op = new DelayedRaftRequest(future, maxWaitTimeMs)
+    purgatory.tryCompleteElseWatch(op, Seq(key))
+    op.isCompletable.set(true)
+  }
+
+  override def completeAll(value: Unit): Unit = {
+    purgatory.checkAndComplete(key)
+  }
+
+  override def numWaiting(): Int = {
+    purgatory.numDelayed
+  }
+}
+
+class DelayedRaftRequest(future: CompletableFuture[Unit], delayMs: Long)
+  extends DelayedOperation(delayMs) {
+
+  val isCompletable = new AtomicBoolean(false)
+  val isExpired = new AtomicBoolean(false)
```

Comment:
Not exactly. The `isCompletable` field is kind of a hack to prevent the operation from getting completed immediately in `await` when we call `tryCompleteElseWatch`. It's similar to the `heartbeatSatisfied` flag in `MemberMetadata`. The problem is that the simpler `RequestPurgatory` interface is not a 1-1 semantic match with `DelayedOperationPurgatory`.